statter_test.go

```diff
@@ -0,0 +1,417 @@
+	m.On("Gauge", "test_min_ms", 5.0, [][2]string{{"tag", "test"}}).Once()
+	m.On("Gauge", "test_max_ms", 45.0, [][2]string{{"tag", "test"}}).Once()
+	m.On("Gauge", "test_10p_ms", 5.0, [][2]string{{"tag", "test"}}).Once()
+	m.On("Gauge", "test_90p_ms", 32.0, [][2]string{{"tag", "test"}}).Once()
+
+	values := []int{10, 20, 10, 30, 20, 11, 12, 32, 45, 9, 5, 5, 5, 10, 23, 8}
+
+	stats := statter.New(m, time.Second)
+
+	timing := stats.Timing("test", tags.Str("tag", "test"))
+	for _, v := range values {
+		timing.Observe(time.Duration(v) * time.Millisecond)
+	}
+
+	err := stats.Close()
+	assert.NoError(t, err)
+
+	m.AssertExpectations(t)
+}
+
+func TestStatter_TimingAggregatedSwapsSamples(t *testing.T) {
+	m := newWaitingReporter()
+	m.On("Counter", "test_count", int64(1), [][2]string{{"tag", "test"}}).Twice()
+	m.On("Gauge", "test_sum_ms", 10.0, [][2]string{{"tag", "test"}}).Twice()
+	m.On("Gauge", "test_mean_ms", 10.0, [][2]string{{"tag", "test"}}).Twice()
+	m.On("Gauge", "test_stddev_ms", 0.0, [][2]string{{"tag", "test"}}).Twice()
+	m.On("Gauge", "test_min_ms", 10.0, [][2]string{{"tag", "test"}}).Twice()
+	m.On("Gauge", "test_max_ms", 10.0, [][2]string{{"tag", "test"}}).Twice()
+	m.On("Gauge", "test_10p_ms", 10.0, [][2]string{{"tag", "test"}}).Twice()
+	m.On("Gauge", "test_90p_ms", 10.0, [][2]string{{"tag", "test"}}).Twice()
+
+	stats := statter.New(m, time.Millisecond)
+
+	stats.Timing("test", tags.Str("tag", "test")).Observe(10 * time.Millisecond)
+
+	select {
+	case <-time.After(time.Second):
+		assert.FailNow(t, "expected call to reporter timed out")
+	case <-m.Ch():
+	}
+	m.Reset()
+
+	stats.Timing("test", tags.Str("tag", "test")).Observe(10 * time.Millisecond)
+
+	select {
+	case <-time.After(time.Second):
+		assert.FailNow(t, "expected call to reporter timed out")
+	case <-m.Ch():
+	}
+
+	err := stats.Close()
+	assert.NoError(t, err)
+
+	m.AssertExpectations(t)
+}
+
+func TestStatter_AggregatedCallsNothingIfNoValues(t *testing.T) {
+	m := &mockSimpleReporter{}
+
+	stats := statter.New(m, time.Second)
+
+	err := stats.Close()
+	assert.NoError(t, err)
+
+	m.AssertNotCalled(t, "Counter", mock.Anything)
+	m.AssertNotCalled(t, "Gauge", mock.Anything)
+}
+
+func TestStatter_TimingReturnsIdenticalCounter(t *testing.T) {
+	stats := statter.New(statter.NullReporter, time.Second)
+	t.Cleanup(func() { _ = stats.Close() })
+
+	t1 := stats.Timing("test", tags.Str("tag", "test"))
+
+	t2 := stats.Timing("test", tags.Str("tag", "test"))
+
+	assert.Same(t, t1, t2)
+}
+
+func TestStatter_CloseFromSubStatterFails(t *testing.T) {
+	stats := statter.New(statter.NullReporter, time.Second).With("prefix", tags.Str("base", "val"))
+
+	err := stats.Close()
+
+	assert.Error(t, err)
+}
+
+func TestNullReporter(t *testing.T) {
+	assert.Implements(t, (*statter.Reporter)(nil), statter.NullReporter)
+	assert.Implements(t, (*statter.HistogramReporter)(nil), statter.NullReporter)
+	assert.Implements(t, (*statter.TimingReporter)(nil), statter.NullReporter)
+}
```

Comment:
This is usually not done in unit tests but rather with a static assertion that prevents the code from even compiling if it's not satisfied. Any reason for doing it this way instead? Is it more of a soft requirement?

Reply:
There is a general argument as to which approach is better. So go for the compile method, others argue what you are doing there is actually a test and therefore should be in the tests. I think it ends up being personal preference 😄